tests/test_sessions.py

```diff
@@ -0,0 +1,161 @@
+        SELECT * FROM bayesdb_session_entries ORDER BY id;
+    '''))
+
+def _basic_test_trace(executor):
+
+    # a new session is automatically initialized with id 1
+    assert get_num_sessions(executor) == 1
+
+    # the above select query counts should become one or more entries
+    num_entries = int(executor('''
+        SELECT COUNT(*) FROM bayesdb_session_entries;
+    ''').next()[0])
+    assert num_entries > 0
+
+    # entries are ordered starting from 1
+    for id, entry in enumerate(get_entries(executor)):
+        assert get_session(entry) == 1
+        assert get_id(entry) == id + 1
+
+def test_sessions_basic_bql():
+    bdb, t = make_bdb_with_sessions()
+    _basic_test_trace(bdb.execute)
+
+def test_sessions_basic_sql():
+    bdb, t = make_bdb_with_sessions()
+    _basic_test_trace(bdb.sql_execute)
+
+def _simple_bql_query(bdb):
+    bdb.execute('''
+        SELECT COUNT(*) FROM bayesdb_session;
+    ''')
+
+def test_sessions_session_id_and_clear_sessions():
+    bdb, t = make_bdb_with_sessions()
+    _simple_bql_query(bdb)
+
+    # create two more sessions
+    t._start_new_session()
+    t._start_new_session()
+    assert t.current_session_id() == 3
+    assert get_num_sessions(bdb.execute) == 3
+
+    # there should now be one session (the current session)
+    t.clear_all_sessions()
+    assert t.current_session_id() == 1
+    assert get_num_sessions(bdb.execute) == 1
+
+    # the entry ids in the current session should start from 1
+    assert min(map(get_id, get_entries(bdb.execute))) == 1
+
+def test_sessions_start_stop():
+    bdb = make_bdb()
+
+    # the session table exists but there should be no entries before we
+    # register the session tracer
+    assert get_num_sessions(bdb.execute) == 0
+    _simple_bql_query(bdb)
+    assert get_num_entries(bdb.execute) == 0
+
+    # registering the tracer starts recording of sessions
+    t = ss.SessionTracer(bdb)
+    _simple_bql_query(bdb)
+    num = get_num_entries(bdb.execute)
+    assert num > 0
+
+    # stopping the tracer
+    t.stop_saving_sessions()
+    _simple_bql_query(bdb)
+    assert get_num_entries(bdb.execute) == num
+
+    # restarting the tracer
+    t.start_saving_sessions()
+    _simple_bql_query(bdb)
+    assert get_num_entries(bdb.execute) > num
+
+def test_sessions_json_dump():
+    bdb, t = make_bdb_with_sessions()
+    _simple_bql_query(bdb)
+    t.stop_saving_sessions()
+    json_str = t.dump_current_session_as_json()
+    entries = json.loads(json_str)
+    assert len(entries) == get_num_entries(bdb.execute)
+
+def test_sessions_unfinished_entry():
+    bdb, t = make_bdb_with_sessions()
+    bql = 'SELECT * FROM nonexistent_table'
+    try:
+        bdb.execute(bql)
+        assert False
+    except sqlite3.OperationalError:
+        t._start_new_session()
+        # check for unfinished queries in the previous session
+        assert t._check_unfinished_entries() > 0
+
+@pytest.mark.skipif(True, reason="Not test network session upload")
```

Comment:
I would actually love to run this test, as part of the usual suite. In part because I'd love to know when the receiver goes down for any reason.